### ISSUE TITLE
mem: StridePrefetcherHashedSetAssociative doesn't work

### ISSUE THREAD

**Issue by erin-le** [gem5/gem5#1371]
**Describe the bug**
When the prefetcher of a cache is set to the StridePrefetcherHashedSetAssociative prefetcher, it causes the simulation to crash.

**Affects version**
gem5 develop for 24.1, 97f6f3c4dafec239258be82a0e8d6eb23b741ec7

**gem5 Modifications**
To more easily change between prefetchers in the configuration script, I added a cache hierarchy based on `src/python/gem5/components/cachehierarchies/classic/private_l1_cache_hierarchy.py` and modified the Sconscript file in src/python to include it.

**To Reproduce**
Steps to reproduce the behavior. Please assume starting from a clean repository:

1. Add the following cache hierarchy to `src/python/gem5/components/cachehierarchies/classic`:

```python
# Copyright (c) 2021 The Regents of the University of California
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met: redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer;
# redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution;
# neither the name of the copyright holders nor the names of its
# contributors may be used to endorse or promote products derived from
# this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

from typing import Optional

from m5.objects import (
    BadAddr,
    BaseXBar,
    Cache,
    Port,
    SystemXBar,
    BasePrefetcher, MultiPrefetcher, QueuedPrefetcher, StridePrefetcherHashedSetAssociative, StridePrefetcher, TaggedPrefetcher, IndirectMemoryPrefetcher, SignaturePathPrefetcher, SignaturePathPrefetcherV2, AMPMPrefetcher, DCPTPrefetcher, IrregularStreamBufferPrefetcher, SlimAMPMPrefetcher, BOPPrefetcher, SBOOEPrefetcher, STeMSPrefetcher, PIFPrefetcher
)

from ....isas import ISA
from ....utils.override import *
from ...boards.abstract_board import AbstractBoard
from ..abstract_cache_hierarchy import AbstractCacheHierarchy
from .abstract_classic_cache_hierarchy import AbstractClassicCacheHierarchy
from .caches.l1dcache import L1DCache
from .caches.l1icache import L1ICache
from .caches.mmu_cache import MMUCache


class PrivateL1CacheHierarchy(AbstractClassicCacheHierarchy):
    """
    A cache setup where each core has a private L1 data and instruction Cache.
    """

    def _get_default_membus(self) -> SystemXBar:
        """
        A method used to obtain the default memory bus of 64 bit in width for
        the PrivateL1CacheHierarchy.

        :returns: The default memory bus for the PrivateL1PrivateL2
                  CacheHierarchy.
        """
        membus = SystemXBar(width=64)
        membus.badaddr_responder = BadAddr()
        membus.default = membus.badaddr_responder.pio
        return membus

    def __init__(
        self,
        l1d_size: str,
        l1i_size: str,
        prefetcher = StridePrefetcher,
        membus: Optional[BaseXBar] = None,
    ) -> None:
        """
        :param l1d_size: The size of the L1 Data Cache (e.g., "32kB").

        :param  l1i_size: The size of the L1 Instruction Cache (e.g., "32kB").

        :param membus: The memory bus. This parameter is optional parameter and
                       will default to a 64 bit width SystemXBar is not
                       specified.
        """

        AbstractClassicCacheHierarchy.__init__(self=self)
        self.membus = membus if membus else self._get_default_membus()
        self._l1d_size = l1d_size
        self._l1i_size = l1i_size
        self._prefetcher = prefetcher

    @overrides(AbstractClassicCacheHierarchy)
    def get_mem_side_port(self) -> Port:
        return self.membus.mem_side_ports

    @overrides(AbstractClassicCacheHierarchy)
    def get_cpu_side_port(self) -> Port:
        return self.membus.cpu_side_ports

    @overrides(AbstractCacheHierarchy)
    def incorporate_cache(self, board: AbstractBoard) -> None:
        # Set up the system port for functional access from the simulator.
        board.connect_system_port(self.membus.cpu_side_ports)

        for _, port in board.get_memory().get_mem_ports():
            self.membus.mem_side_ports = port

        self.l1icaches = [
            L1ICache(size=self._l1i_size, PrefetcherCls=self._prefetcher)
            for i in range(board.get_processor().get_num_cores())
        ]

        self.l1dcaches = [
            L1DCache(size=self._l1d_size, PrefetcherCls=self._prefetcher)
            for i in range(board.get_processor().get_num_cores())
        ]
        # ITLB Page walk caches
        self.iptw_caches = [
            MMUCache(size="8KiB")
            for _ in range(board.get_processor().get_num_cores())
        ]
        # DTLB Page walk caches
        self.dptw_caches = [
            MMUCache(size="8KiB")
            for _ in range(board.get_processor().get_num_cores())
        ]

        if board.has_coherent_io():
            self._setup_io_cache(board)

        for i, cpu in enumerate(board.get_processor().get_cores()):
            cpu.connect_icache(self.l1icaches[i].cpu_side)
            cpu.connect_dcache(self.l1dcaches[i].cpu_side)

            self.l1icaches[i].mem_side = self.membus.cpu_side_ports
            self.l1dcaches[i].mem_side = self.membus.cpu_side_ports

            self.iptw_caches[i].mem_side = self.membus.cpu_side_ports
            self.dptw_caches[i].mem_side = self.membus.cpu_side_ports

            cpu.connect_walker_ports(
                self.iptw_caches[i].cpu_side, self.dptw_caches[i].cpu_side
            )

            if board.get_processor().get_isa() == ISA.X86:
                int_req_port = self.membus.mem_side_ports
                int_resp_port = self.membus.cpu_side_ports
                cpu.connect_interrupt(int_req_port, int_resp_port)
            else:
                cpu.connect_interrupt()

    def _setup_io_cache(self, board: AbstractBoard) -> None:
        """Create a cache for coherent I/O connections"""
        self.iocache = Cache(
            assoc=8,
            tag_latency=50,
            data_latency=50,
            response_latency=50,
            mshrs=20,
            size="1kB",
            tgts_per_mshr=12,
            addr_ranges=board.mem_ranges,
        )
        self.iocache.mem_side = self.membus.cpu_side_ports
        self.iocache.cpu_side = board.get_mem_side_coherent_io_port()
```
2. Add the following to `src/python/SConscript`:
```python
PySource('gem5.components.cachehierarchies.classic',
    'gem5/components/cachehierarchies/classic/private_l1_cache_hierarchy_prefetch.py')
```
3. Compile gem5 with command `build/ALL/gem5.fast -j [num of cores]`
4. Execute the simulation with the following command and configuration script:

Command (assumes you named the config file arm-hello-prefetch.py and put it in the top level of the gem5 folder):
```bash
./build/ALL/gem5.fast arm-hello-prefetch.py --prefetcher=StridePrefetcherHashedSetAssociative
```
Configuration script:
```python
# Copyright (c) 2021 The Regents of the University of California
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met: redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer;
# redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution;
# neither the name of the copyright holders nor the names of its
# contributors may be used to endorse or promote products derived from
# this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

"""
This gem5 configuation script creates a simple board to run an ARM
"hello world" binary.

This is setup is the close to the simplest setup possible using the gem5
library. It does not contain any kind of caching, IO, or any non-essential
components.

"""

from m5.objects import (BasePrefetcher, MultiPrefetcher, QueuedPrefetcher, StridePrefetcherHashedSetAssociative, StridePrefetcher, TaggedPrefetcher, IndirectMemoryPrefetcher, SignaturePathPrefetcher, SignaturePathPrefetcherV2, AMPMPrefetcher, DCPTPrefetcher, IrregularStreamBufferPrefetcher, SlimAMPMPrefetcher, BOPPrefetcher, SBOOEPrefetcher, STeMSPrefetcher, PIFPrefetcher)

from gem5.components.boards.simple_board import SimpleBoard
from gem5.components.cachehierarchies.classic.private_l1_cache_hierarchy_prefetch import PrivateL1CacheHierarchy
from gem5.components.memory import SingleChannelDDR3_1600
from gem5.components.processors.cpu_types import CPUTypes
from gem5.components.processors.simple_processor import SimpleProcessor
from gem5.isas import ISA
from gem5.resources.resource import obtain_resource
from gem5.simulate.simulator import Simulator
from gem5.utils.requires import requires
import argparse

parser = argparse.ArgumentParser()

parser.add_argument(
    "--prefetcher",
    type=str,
    help="Enter the prefetcher you would like to use."
    )

args = parser.parse_args()

# This check ensures the gem5 binary is compiled to the ARM ISA target. If not,
# an exception will be thrown.
requires(isa_required=ISA.ARM)

str_to_prefetch_class = {"BasePrefetcher":BasePrefetcher, "MultiPrefetcher":MultiPrefetcher, "QueuedPrefetcher":QueuedPrefetcher, "StridePrefetcherHashedSetAssociative":StridePrefetcherHashedSetAssociative, "StridePrefetcher":StridePrefetcher, "TaggedPrefetcher":TaggedPrefetcher, "IndirectMemoryPrefetcher":IndirectMemoryPrefetcher, "SignaturePathPrefetcher":SignaturePathPrefetcher, "SignaturePathPrefetcherV2":SignaturePathPrefetcherV2, "AMPMPrefetcher":AMPMPrefetcher, "DCPTPrefetcher":DCPTPrefetcher, "IrregularStreamBufferPrefetcher":IrregularStreamBufferPrefetcher, "SlimAMPMPrefetcher":SlimAMPMPrefetcher, "BOPPrefetcher":BOPPrefetcher, "SBOOEPrefetcher":SBOOEPrefetcher, "STeMSPrefetcher":STeMSPrefetcher, "PIFPrefetcher":PIFPrefetcher}

# In this setup we don't have a cache. `NoCache` can be used for such setups.
cache_hierarchy = PrivateL1CacheHierarchy(l1d_size="32kB", l1i_size="32kB", prefetcher=str_to_prefetch_class[args.prefetcher])

# We use a single channel DDR3_1600 memory system
memory = SingleChannelDDR3_1600(size="32MB")

# We use a simple Timing processor with one core.
processor = SimpleProcessor(cpu_type=CPUTypes.TIMING, isa=ISA.ARM, num_cores=1)

# The gem5 library simble board which can be used to run simple SE-mode
# simulations.
board = SimpleBoard(
    clk_freq="3GHz",
    processor=processor,
    memory=memory,
    cache_hierarchy=cache_hierarchy,
)

# Here we set the workload. In this case we want to run a simple "Hello World!"
# program compiled to the ARM ISA. The `Resource` class will automatically
# download the binary from the gem5 Resources cloud bucket if it's not already
# present.
board.set_se_binary_workload(
    # The `Resource` class reads the `resources.json` file from the gem5
    # resources repository:
    # https://github.com/gem5/gem5-resources.
    # Any resource specified in this file will be automatically retrieved.
    # At the time of writing, this file is a WIP and does not contain all
    # resources. Jira ticket: https://gem5.atlassian.net/browse/GEM5-1096
    obtain_resource("arm-hello64-static", resource_version="1.0.0")
)

# Lastly we run the simulation.
simulator = Simulator(board=board)
simulator.run()

print(
    "Exiting @ tick {} because {}.".format(
        simulator.get_current_tick(), simulator.get_last_exit_event_cause()
    )
)

```

**Terminal Output**

```shell
info: Using default config
TypeError: BasePrefetcher.__init__() takes 1 positional argument but 2 were given
Error setting param L1ICache.prefetcher to <orphan StridePrefetcherHashedSetAssociative>


At:
  src/python/m5/params.py(224): convert
  src/python/m5/SimObject.py(888): __setattr__
  src/python/gem5/components/cachehierarchies/classic/caches/l1icache.py(67): __init__
  src/python/gem5/components/cachehierarchies/classic/private_l1_cache_hierarchy_prefetch.py(106): <listcomp>
  src/python/gem5/components/cachehierarchies/classic/private_l1_cache_hierarchy_prefetch.py(105): incorporate_cache
  src/python/gem5/components/boards/abstract_board.py(373): _connect_things
  src/python/gem5/components/boards/abstract_board.py(392): _pre_instantiate
  src/python/gem5/simulate/simulator.py(641): _instantiate
  src/python/gem5/simulate/simulator.py(722): run
  /home/bees/gem5-2nd-worktree/configs/example/gem5_library/running-gem5-resources/arm-hello-prefetch.py(107): <module>
  src/python/m5/main.py(669): main
```

**Expected behavior**
For the simulation to run without crashing. With the included config script, the simulation should run, print "Hello World!", then exit.

**Host Operating System**
Ubuntu 22.04

**Host ISA**
ARM

**Comment by AbhishekUoR** [comment]
Just set to StridePrefetcher

**Comment by BobbyRBruce** [comment]
Fixed by #1374
